### PR TITLE
Neo VM: measure transaction execution

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr adds a measurement tool to the neo vm. When enabled, it can measure the exact time cost to execute every single opcode in every transaction, including the preexecution time, execution time, postexecutiontime, and the gas efficiency.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
@cschuchardt88 for me, its more like a local node measurement, not necessarily to be a final solution, just tools that i used in the past few days. If you can open another pr with a better solution, welcome.

**Comment by Jim8y** [comment]
BTW, time is used to measure the gas efficiency among opcodes. not really used to measure `how long` it runs, but which runs longer.

**Comment by cschuchardt88** [comment]
If we keep doing it wrong. It will never get better and take someone longer to build and fix existing code. Each line added is another line someone will have to change.

**Comment by Jim8y** [comment]
> If we keep doing it wrong. It will never get better and take someone longer to build and fix existing code. Each line added is another line someone will have to change.

First of all, its just my pr, not a merge, you can disagree. Second, i've been using this to benchmark the vm, not a good solution maybe, but works at least for me, you can propose your solution, i wont against that. Third, what is better and what is worse, leave it there and pretent nothing is wrong, or propose a `wrong` as you said, then we discuss it and come up with a `good` solution?

Now, you think it is not a good idea to use time to measure, then what is your proposal? How do we know the best and worst case of the vm? how do we know that each opcode is properly priced, but their combination will also work correctly?

We need a solution, doing it wrong, my bad, but can we get one?

**Comment by cschuchardt88** [comment]
Prices should be a combination of `CPU` and `latency` with a sample rate.
- `latency` will measure the time it takes to complete an `IO` operation in two parts `Read`, `Write`. This includes `Memory`, `Disk` and `Network`.
- `CPU` will measure the amount of CPU it takes when done processing a method. Sampled over time to get the average peek.
- `Amount` this one is straight forward. Measure the amount of `Memory` and `Disk Space` a operation uses.

If you measure these two things it pretty clear picture what is being used. Time wouldn't matter at all and the reason for this, is because faster computer running nothing but the application will run faster than a computer loaded with tons of applications. Time measurement is only used to measure how long a process waits on a response from `a person`, `network` or `another process`. For time to work you a lot of data. Measuring a method that takes milliseconds doesn't say anything. The thread that starts the process could be anywhere in the queue of threads in your operating system. One run could put it at position `7` and the next run, put it at position `40` in the thread queue.

**Comment by shargon** [comment]
> If we keep doing it wrong. It will never get better and take someone longer to build and fix existing code. Each line added is another line someone will have to change.

I think that we can do this with Visual studio profiler

**Comment by Jim8y** [comment]
> > If we keep doing it wrong. It will never get better and take someone longer to build and fix existing code. Each line added is another line someone will have to change.
> 
> I think that we can do this with Visual studio profiler

Its different, i want precise measurement on every single OpCode, i want to evaluate the Gas efficiency of every OpCode on different situations, for instance, when it may takes a very long time because of GC or reference clean.

File: src/Neo.VM/OpCode.Price.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using System.Collections.Generic;
+namespace Neo.VM
```
Comment (shargon) [inline src/Neo.VM/OpCode.Price.cs:13]:
```suggestion
using System.Collections.Generic;

namespace Neo.VM
```

File: src/Neo.VM/ExecutionEngine.cs [hunk 2]
```diff
@@ -133,12 +133,19 @@ public virtual void Dispose()
         /// Start execution of the VM.
         /// </summary>
         /// <returns></returns>
-        public virtual VMState Execute()
+        public virtual VMState Execute(string txHash = "")
```
[the remark this answers is not in the document]
Comment (shargon) [reply src/Neo.VM/ExecutionEngine.cs]: @Jim8y please take a look to this refactor

**Comment by Jim8y** [comment]
sure, but please review the `waiting for review` pr first, this one will be updated later.

**Comment by cschuchardt88** [comment]
> We should use `IDiagnostic` in Neo.SmartContract

No, `IApplicationEngineProvider` is better.

**Pull request closed without merging**